importer/av_json_data_importer.py

```diff
@@ -0,0 +1,107 @@
+from database.database_data import DatabaseData
+from datetime import datetime
+import importer.json_key_handler as json_kh
+import json
+
+AV_ALIASES = {
+    "1. open": "open",
+    "2. high": "high",
+    "3. low": "low",
+    "4. close": "close",
+    "5. volume": "volume"
+}
+
+METADATA_KEY = "metadata"
+DATETIME_KEY = "datetime"
+
+
+class AVJSONDataImporter(JSONDataImporter):
+    '''
+    Imports Alpha Vantage historycal data into the database
+    '''
+
+    def from_contents(self, json_file_contents: dict):
+        '''
+        Imports data given a json document content.
+
+        Parameters:
+            json_file_contents : dict
+                A dictionary of key-values, should be loaded using json.load(filepath) from a file or DataFrame.__to_json(orient="table", indent=4)
+                Expects a datetime key and a dict value.
+        '''
+        new_data = self._rename_keys(json_file_contents)
+        new_data = self._add_timestamp(new_data)
+        atoms = self._extract_atoms(new_data)
+        atoms = self.__add_metadata_to_atoms(new_data[METADATA_KEY], atoms)
+        atoms = self.__fix_atoms_datetime(atoms)
+        self.database.write(DatabaseData("atoms_b", atoms))
+
+    def _add_timestamp(self, data_dict: dict):
+        '''
+        Copies the timestamp key inside each atom.
+
+        Parameters:
+            data_dict : dict
+                A dictionary of key-values pairs. Should be:
+                {datetime : dict, datetime : dict, ...}
+                For each dict item, will attempt to add the "datetime" field, using the item's key as the new field's value.
+                Ignores item with the key METADATA_KEY, if any.
+        Returns:
+            The same dictionary it received, after modifying it.
+        '''
+        for key, value in data_dict.items():
+            if key == METADATA_KEY or type(value) != dict:
+                continue
+            value[DATETIME_KEY] = key
+        return data_dict
+
+    def _extract_atoms(self, data_dict: dict):
+        '''
+        Takes the atoms list out of the data dictionary.
+
+        Parameters:
+            data_dict : dict
+                A dictionary. Any fields with the METADATA_KEY will be ignored.
+        Returns:
+            The list of values in the dictionary except METADATA_KEY
+        '''
+        atoms = list(data_dict.values())
+        if METADATA_KEY in data_dict.keys():
+            atoms.remove(data_dict[METADATA_KEY])
+        return atoms
+
+    def __fix_atoms_datetime(self, atoms: list):
+        for atom in atoms:
+            atom[DATETIME_KEY] = datetime.strptime(
+                atom[DATETIME_KEY], "%Y-%m-%d %H:%M:%S"
+            ).strftime("%Y-%m-%d %H:%M:%S.%f")[:-3]
+        return atoms
+
+    def __add_metadata_to_atoms(self, metadata: dict, atoms: list):
+        '''
+        Adds each metadata field to all dictionary elements of the given list
```

Comment:
A little confusing, try:
"Adds each metadata field to all ~dictionary~ atoms of the given list"